demo/parser_demo.cpp

```diff
@@ -5,140 +5,216 @@
+    {
+        if (node == nullptr)
+        {
             fmt::print("{:|>{}}- [ArrayType]> null\n", "", depth);
-        } else {
+        }
+        else
+        {
             fmt::print("{:|>{}}- [ArrayType]>\n", "", depth);
             depth++;
             node->length->accept(*this);
             node->elementType->accept(*this);
             depth--;
         }
     };
-    void visit(ast::RecordType *node) override {
-        if (node == nullptr) {
+    void visit(ast::RecordType *node) override
+    {
+        if (node == nullptr)
+        {
             fmt::print("{:|>{}}- [RecordType]> null\n", "", depth);
-        } else {
+        }
+        else
+        {
             fmt::print("{:|>{}}- [RecordType]>\n", "", depth);
             depth++;
-            for (auto field : node->fields) {
+            for (auto field : node->fields)
+            {
                 field->accept(*this);
             }
             depth--;
         }
     };
-    void visit(ast::Variable *node) override {
+    void visit(ast::Variable *node) override{
 
     };
-    void visit(ast::Body *node) override {
+    void visit(ast::Body *node) override{
 
     };
-    void visit(ast::Statement *node) override {
+    void visit(ast::Statement *node) override{
 
     };
-    void visit(ast::Assignment *node) override {
+    void visit(ast::Assignment *node) override{
 
     };
-    void visit(ast::RoutineCall *node) override {
+    void visit(ast::RoutineCall *node) override{
 
     };
-    void visit(ast::WhileLoop *node) override {
-
+    void visit(ast::WhileLoop *node) override
+    {
+        if (node == nullptr)
+        {
+            fmt::print("{:|>{}}- [WhileLoop]> null\n", "", depth);
+        }
+        else
+        {
+            fmt::print("{:|>{}}- [WhileLoop]>\n", "", depth);
+            depth++;
+            node->condition->accept(*this);
+            node->body->accept(*this);
+            depth--;
+        }
     };
-    void visit(ast::ForLoop *node) override {
-
+    void visit(ast::ForLoop *node) override
+    {
+        if (node == nullptr)
+        {
+            fmt::print("{:|>{}}- [ForLoop]> null\n", "", depth);
+        }
+        else
+        {
+            fmt::print("{:|>{}}- [ForLoop]>\n", "", depth);
+            depth++;
+            // node->loopVar->accept(*this);
+            node->rangeFrom->accept(*this);
+            node->rangeTo->accept(*this);
+            node->body->accept(*this);
+            depth--;
+        }
     };
-    void visit(ast::IfStatement *node) override {
+    void visit(ast::IfStatement *node) override
+    {
 
+        if (node == nullptr)
+        {
+            fmt::print("{:|>{}}- [IfStatement]> null\n", "", depth);
+        }
+        else
+        {
+            fmt::print("{:|>{}}- [IfStatement]>\n", "", depth);
+            depth++;
+            node->condition->accept(*this);
+            node->ifBody->accept(*this);
+            node->elseBody->accept(*this);
```

Comment:
At least this one can be `null` for sure, since we don't always have to have a body (and an empty body is not like no body at all)